### PR TITLE
Share Hexagon runtime in offload pipelines

### PR DESCRIPTION
**Pull request by dsharletg**
This change makes it so offload pipelines use Target::NoRuntime. The runtime is instead compiled into a separate module. This is a breaking change to the Hexagon remote runtime.

This change will enable us to move parts of the runtime currently shipped in the Hexagon remote runtime, which is an improvement because these are harder to update/fix bugs in, and causes a gap in features/support for offload vs. standalone pipelines.

### CODE REVIEW

File: src/LLVM_Runtime_Linker.cpp [hunk 1]
```diff
@@ -726,7 +732,11 @@ std::unique_ptr<llvm::Module> get_initial_module_for_target(Target t, llvm::LLVM
             modules.push_back(get_initmod_gpu_device_selection(c, bits_64, debug));
             modules.push_back(get_initmod_tracing(c, bits_64, debug));
             modules.push_back(get_initmod_write_debug_image(c, bits_64, debug));
-            modules.push_back(get_initmod_cache(c, bits_64, debug));
+            if (t.arch != Target::Hexagon) {
+                // TODO: Support this module in the Hexagon backend,
```
Comment (steven-johnson) [inline src/LLVM_Runtime_Linker.cpp]: Does this need to be completed before this can land?

**Comment by abadams** [comment]
Does using -Wl,--whole-archive when turning a static runtime into a shared
runtime not export the symbols you need?

**Comment by abadams** [comment]
Oh, I forgot that we were writing our own shared object with custom code. Hrm... should exporting symbols be the default behavior of write_shared_object (or a parameter to it), or should it be a target flag...

File: src/LLVM_Runtime_Linker.cpp (hunk 1 above)
Comment (dsharletg) [inline src/LLVM_Runtime_Linker.cpp]: I don't think so. This never would have worked before, it just used to get stripped from the runtime, now it doesn't, so we hit the issue.

**Comment by dsharletg** [comment]
I think Target::SharedRuntime is useful even outside this context. Even if it were a parameter to write_shared_object, it would not be easy to plumb it through (the best way to do that would be ... a target flag).

I also think it would be good to have a shared runtime have *non* weak symbols, so you actually get linker errors if you try to have two of them.

File: src/runtime/hexagon_host.cpp [hunk 2]
```diff
@@ -202,27 +203,51 @@ WEAK int halide_hexagon_initialize_kernels(void *user_context, void **state_ptr,
     // created/inserted here.
     ScopedMutexLock lock(&thread_lock);
 
+    // Initialize the runtime, if necessary.
+    if (!shared_runtime) {
+        debug(user_context) << "    Initializing shared runtime\n";
+        const char soname[] = "libshared_runtime.so";
```
Comment (ronlieb) [inline src/runtime/hexagon_host.cpp]: would it be good to have a slightly halide centric name for this library  "libshared_runtime.so" ?

Comment (dsharletg) [inline src/runtime/hexagon_host.cpp]: Keep in mind these never hit the file system so it shouldn't be a big deal. But yeah, I will change it.

**Comment by ronlieb** [comment]
LGTM

**Comment by abadams** [comment]
lgtm

File: Makefile [hunk 3]
```diff
@@ -548,6 +548,7 @@ RUNTIME_CPP_COMPONENTS = \
   float16_t \
   gcd_thread_pool \
   gpu_device_selection \
+  hexagon_cpu_features \
```
Comment (pranavb-ca) [inline Makefile]: What is this needed for?

File: src/HexagonOffload.cpp [hunk 4]
```diff
@@ -899,9 +909,13 @@ Buffer<uint8_t> compile_module_to_hexagon_shared_object(const Module &device_cod
     }
 
     // Link into a shared object.
+    std::string soname = "lib" + device_code.name() + ".so";
     Elf::HexagonLinker linker(device_code.target());
-    std::vector<std::string> dependencies = { "libhalide_hexagon_remote_skel.so" };
-    std::vector<char> shared_object = obj->write_shared_object(&linker, dependencies);
+    std::vector<std::string> dependencies = {
+        "lib" + runtime_module_name + ".so",
```
Comment (pranavb-ca) [inline src/HexagonOffload.cpp]: Should the shared_runtime.so have a dependency on itself?

File: Makefile (hunk 3 above)
Comment (dsharletg) [inline Makefile]: Turns out it's not anymore! It was needed when the runtime wasn't declared weak, another runtime module depended on this being available.

File: src/HexagonOffload.cpp (hunk 4 above)
Comment (dsharletg) [inline src/HexagonOffload.cpp]: This is a little dumb but doesn't seem to hurt anything, it *does* technically depend on itself :)

**Comment by pranavb-ca** [comment]
Tried to run HelloHexagon on this branch
rm -rf ./bin
g++ -std=c++11 -I ../../include/ -I ../../tools/ -I ../../apps/support/ -g -fno-rtti pipeline.cpp ../../lib/libHalide.a ../../tools/GenGen.cpp -o bin/pipeline  -ldl -lpthread -lz -lrt -ldl -lpthread -lz -lm
bin/pipeline -o ./bin/arm-64-android -e o,h -f pipeline_cpu target=arm-64-android
bin/pipeline -o ./bin/arm-64-android -e o,h -f pipeline_hvx64 target=arm-64-android-hvx_64
Internal error at /local/mnt2/dev/sources/Halide/debug/Halide/src/HexagonOffload.cpp:464 triggered by user code at /local/mnt2/dev/sources/Halide/debug/Halide/src/Elf.cpp:810:
Unhandled relocation type 70
make: *** [bin/arm-64-android/pipeline_hvx64.o] Aborted
rm bin/arm-64-android/pipeline_cpu.o

Are you seeing this at all?

**Comment by dsharletg** [comment]
I just tried it on an SD835 device (uses dlopenbuf), it worked fine. I will try it on an SD820 (uses mmap/mprotect) when I get in tomorrow. The mmap/mprotect codepath does work on the simulator, so I'd be a bit surprised if it doesn't work given that everything else works on device.

Do you still see the same issue when running on the simulator? The error you are seeing (in the linker for the Hexagon code) shouldn't actually be different for simulator vs. device.

**Comment by dsharletg** [comment]
This is working for me on an SD820 device also.

Do you have any other changes in your branch locally (thread pool stuff?)? Or maybe a local LLVM change that is generating different code?

Either way, we should probably just add support for that relocation. Does the following patch resolve the issue for you?

diff --git a/src/HexagonOffload.cpp b/src/HexagonOffload.cpp
index 814879b48..4cc2f0dc2 100644
--- a/src/HexagonOffload.cpp
+++ b/src/HexagonOffload.cpp
@@ -455,6 +455,10 @@ void do_relocation(uint32_t fixup_offset, char *fixup_addr, uint32_t type,
         do_reloc(fixup_addr, Word32_X26, intptr_t(G) >> 6, _signed, truncate);
         needs_got_entry = true;
         break;
+    case R_HEX_GOT_16_X:
+        do_reloc(fixup_addr, Word32_U6, intptr_t(G), _signed, truncate);
+        needs_got_entry = true;
+        break;
     case R_HEX_GOT_11_X:
         do_reloc(fixup_addr, Word32_U6, uintptr_t(G), _unsigned, truncate);
         needs_got_entry = true;

**Comment by dsharletg** [comment]
Wait, it seems like there was a stale build issue. I'm seeing a different issue that needs fixing.

**Comment by dsharletg** [comment]
It seems like this is uncovering yet more inexplicable behavior in Hexagon's dlsym. I'm sick of fighting this issue, I'm going to port over what the simulator does, so we can be done with this problem.

**Comment by pranavb-ca** [comment]
I'll try the relocation fix, but the different issue that you said you were seeing, what is that issue? Or should I just wait for your new patch?

**Comment by pranavb-ca** [comment]
The relocation failure was fixed by your patch. This is the problem I am seeing now
Hexagon: halide_hexagon_initialize_kernels (user_context: 0x0, state_ptr: 0x5584bf1c60, *state_ptr: 0x0, code: 0x5584bc7ae0, code_size: 21040)
05-11 00:16:49.760 29873 29873 I halide  : , code: 0x5584bccd40, code_size: 49752)
05-11 00:16:49.760 29873 29873 I halide  :     Initializing shared runtime
05-11 00:16:49.760 29873 29873 I halide  :     halide_remote_load_library(libhalide_shared_runtime.so) ->
05-11 00:16:49.761 29873 29873 I halide  : Unresolved external symbol fileno
05-11 00:16:49.761 29873 29873 I halide  : mmap_dlopen failed
05-11 00:16:49.762 29873 29873 I halide  :         -1
05-11 00:16:49.762 29873 29873 I halide  : Error: Initialization of Hexagon kernels failed

**Comment by dsharletg** [comment]
I've actually found a bunch of issues. I've pushed some commits that should fix some of them:

- hexagon_cpu_features needed after all.
- dlsym not working in strange ways (not new), worked around that by porting the simulator's version of halide_get_symbol.
- The simulator's halide_get_symbol is breaking on device on SD835, because it needs mmap/mprotect, despite those symbols being declared weak. It seems like maybe dlopen on device is not handling weak symbols correctly? I'm working on a workaround.
- dlopenbuf is working on the simulator, but not on device (dlopenbuf failed: RX VA 0x1F100000 outside ELF segment). Note it should be the exact same binary in both places, so something interesting is going on here. I'm going to debug that once I get through the first two issues.

Here's where it is working:
- SD820, mmap_dlopen
- Simulator, mmap_dlopen
- Simulator, dlopenbuf

Not working:
- SD835, dlopenbuf

Not available to test (not supported by the firmware I have on these devices):
- SD820, dlopenbuf
- SD835, mmap_dlopen

**Comment by dsharletg** [comment]
OK, messages crossed, that recent round of commits should have fixed the issue you were seeing with fileno.

**Comment by pranavb-ca** [comment]
Trying your changes now. Thanks for the fixes!

**Comment by dsharletg** [comment]
Hmm, I've fixed the remaining issues that I described before.

I'm now hitting what appears to be a showstopper: it looks like dlopenbuf needs "libhalide_shared_runtime.so" in the DT_NEEDED header. However, it looks like dlopenbuf also *requires* this to be a real file system shared library. So, if I put "libhalide_shared_runtime.so" in the DT_NEEDED header, I get this error, even if I dlopenbuf that library (with the appropriate soname) prior to attempting it:

    dlopenbuf failed: cannot open libhalide_shared_runtime.so

If I remove "libhalide_shared_runtime.so" from the DT_NEEDED header, it can't find the symbols it needs.

Note that this behavior differs between the simulator and on-device... :(

**Comment by dsharletg** [comment]
Hmm, I *might* be wrong about dlopenbuf needing libhalide_shared_runtime.so in the DT_NEEDED header. I'm actually seeing "dlopenbuf failed: RX VA 0x1F100000 outside ELF segment" when I remove it now, not missing symbols. This is weird, it does work on the simulator and with mmap_dlopen, so I don't think the shared object is malformed.

**Comment by dsharletg** [comment]
I just found a workaround for the issue above. Pushed a commit, this is now working on every configuration I can test (SD 820, mmap_dlopen, SD 835, dlopenbuf, simulator mmap_dlopen, simulator dlopenbuf).

**Comment by pranavb-ca** [comment]
Thanks. I'll pull your changes now and try them out

**Comment by ronlieb** [comment]
Pranav asked me to try them on my 8998, and they worked fine.

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

**Comment by dsharletg** [comment]
I just did some more testing of this to consider merging it, when I realized that this is necessarily a breaking change to the runtime. This is not an issue for us, and it was not an issue for anyone else last time we made a breaking change to the runtime...

**Comment by ronlieb** [comment]
Noted, thx

 

I think you should proceed

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

**Pull request merged**